### ISSUE TITLE
Reproducing the numbers given

### ISSUE THREAD

**Issue by prabhuteja12** [yzou2/CRST#22]
Hi, 
I'm trying to run your code to reproduce the results. However, I'm running into some issues. Using the same packages (pytorch, python versions), I'm getting the following the results. 
| Method | Weight | Result |
------------ | ---------- | -------- |
| lr_ent |  0.25 | 44.59% |
| cbst   | | 2.54% |
| mr_weight_kld | 0.1 |  2.86% |

I've used the defaults prescribed in the codebase itself. 

Can you tell me what might be doing wrong?

**Comment by yzou2** [comment]
Did you resize the images of GTA5 to 1052x1914?

**Comment by prabhuteja12** [comment]
hi, yes. they are of that size.

**Comment by yzou2** [comment]
That seems weird to me. If the LRENT works, so does the CBST and MRKLD. We need to localize the bug. Can you show me the source model predictions & pseudo-label of CBST and MRKLD? Also in the retraining process, is the loss normal?

**Comment by prabhuteja12** [comment]
Hi, 
The loss never explodes (no NaN or Inf), but I observe a really weird pattern in the IoU of the test set. 
| Round | CBST | LRENT | MRKLD |
|----------|----------|----------|----------|
| 0 | 3.54 | 40.52 | 3.64 |
| 1 | 33.03 | 43.80 | 32.85 |
| 2 | 2.54 | 44.59 | 44.59 |

There is spike after round-1 that seems to disappear right away. 

Here are the figures you ask for: 

![cbst_aachen_000029_000019_leftImg8bit_color](https://user-images.githubusercontent.com/11191577/79915407-31a96f00-8427-11ea-8df5-9497158c06d0.png)
![lrent_aachen_000029_000019_leftImg8bit_color](https://user-images.githubusercontent.com/11191577/79915409-32420580-8427-11ea-9c6d-1c5e74b33811.png)
![mrkld_aachen_000029_000019_leftImg8bit_color](https://user-images.githubusercontent.com/11191577/79915410-32da9c00-8427-11ea-9fa6-7347d3c998ac.png)

It seems that something is changing at the first iterations, where the road has been flipped, building etc.. I think that might be the cause of the problem.  

Let me know if you need some more details.

**Comment by yzou2** [comment]
Thanks for the details. It seems that the results of round 0 in your table is the numbers after first round of self-training. Can you provide the mIoU score and predictions/pseudo-labels for the source only model? If the source model's predictions are not incorrect, there should be something in evaluation.

And to clarify, the 1,2,3 rows in the figures correspond to 0,1,2 round results?

**Comment by prabhuteja12** [comment]
Hi, 
Sorry that I forgot to give those details. The image I used is "aachen_000029_000019_leftImg8bit_color.png" from the train set. I didn't change the directory structure of your saving code. So, I took '<algo>/<iteration_no>/pred_vis/aachen_000029_000019_leftImg8bit_color.png'  and made the above figures. 

The numbers are post training the network each round of training. I hope this clears it up.

**Comment by yzou2** [comment]
Thanks for the clarification. So I think the source model and initial pseudo-labels are correct. Something bad happened in the retraining part of initial round. Can you show me the loss of initial round and are they reasonable small?

**Comment by yzou2** [comment]
BTW, what's the version of OpenCV? It should be OpenCV 3.2.0.

**Comment by yzou2** [comment]
LRENT has different MRKLD and CBST. So it may be the library called in the dataset function is different than my settings. You may want to check it.

**Comment by prabhuteja12** [comment]
`cv2.__version` is 3.2.0. 

The training logs are as follows 
`2020-04-16 15:04:02,592 Host ###### Start model retraining dataset in round 0! ######
2020-04-16 15:04:07,246 Host iter = 1 of 552.0 completed, loss = 10.13249683380127
2020-04-16 15:04:08,915 Host iter = 2 of 552.0 completed, loss = 7.9210052490234375
2020-04-16 15:04:11,047 Host iter = 3 of 552.0 completed, loss = 6.108864784240723
2020-04-16 15:04:12,967 Host iter = 4 of 552.0 completed, loss = 1.668105959892273
2020-04-16 15:04:14,475 Host iter = 5 of 552.0 completed, loss = 1.0887787342071533
2020-04-16 15:04:15,946 Host iter = 6 of 552.0 completed, loss = 0.7527024149894714
2020-04-16 15:04:17,576 Host iter = 7 of 552.0 completed, loss = 0.46681663393974304
2020-04-16 15:04:19,122 Host iter = 8 of 552.0 completed, loss = 1.0938246250152588
2020-04-16 15:04:20,788 Host iter = 9 of 552.0 completed, loss = 2.2647829055786133
2020-04-16 15:04:22,447 Host iter = 10 of 552.0 completed, loss = 0.9959995150566101
`

The full log for cbst is here 
[self_training_log.txt](https://github.com/yzou2/CRST/files/4513993/self_training_log.txt)

**Comment by yzou2** [comment]
In the initial round of self-training, the loss is reasonable, which should indicate the predictions are reasonably correct. However, the predictions are mostly incorrect after the initial self-training. That's really weird... Can you check the values of the image of '/1/pred/aachen_000029_000019_leftImg8bit_color.png' (you can upload the raw image). If the predications values are correct. Maybe the problem is in the label mapping function

**Comment by prabhuteja12** [comment]
Here is the image:
![aachen_000029_000019_leftImg8bit](https://user-images.githubusercontent.com/11191577/79950311-2b42e380-8477-11ea-9951-23af598e645c.png)

In the code 
``` python
if args.lr_weight_ent == 0.0:
       srctrainset = SrcSTDataSet(args.data_src_dir, src_train_lst, max_iters=src_num_sel,reg_weight=reg_weight_src,data_src=args.data_src, crop_size=input_size,scale=args.random_scale, mirror=args.random_mirror,train_scale=train_scale_src, mean=IMG_MEAN, std=IMG_STD)
       tgttrainset = GTA5StMineDataSet(args.data_tgt_dir, tgt_train_lst, pseudo_root=save_pseudo_label_path, max_iters=tgt_num,reg_weight=reg_weight_tgt,rare_id = rare_id, mine_id=mine_id, mine_chance = mine_chance, crop_size=input_size,scale=args.random_scale,data_src=args.data_src, mirror=args.random_mirror, train_scale=train_scale_tgt, mean=IMG_MEAN, std=IMG_STD)
elif args.lr_weight_ent > 0.0:
       srctrainset = SoftSrcSTDataSet(args.data_src_dir, src_train_lst, max_iters=src_num_sel,reg_weight=reg_weight_src,data_src=args.data_src,num_classes=args.num_classes, crop_size=input_size,scale=args.random_scale, mirror=args.random_mirror, train_scale=train_scale_src, mean=IMG_MEAN, std=IMG_STD)
       tgttrainset = SoftGTA5StMineDataSet(args.data_tgt_dir, tgt_train_lst, pseudo_root=save_pseudo_label_path, max_iters=tgt_num,reg_weight=reg_weight_tgt,rare_id = rare_id, mine_id=mine_id, mine_chance = mine_chance, crop_size=input_size,scale=args.random_scale,data_src=args.data_src,num_classes=args.num_classes, mirror=args.random_mirror, train_scale=train_scale_tgt, mean=IMG_MEAN, std=IMG_STD)
```

In the codes for `GTA5TestDataSet` there doesn't seem to be an explicit label mapper. Is this intentional? Is this done else where in the `crst_seg.py` file itself?

**Comment by yzou2** [comment]
Yes, it is intentional. The scoreupdater can handle the label mapping (The mIoU on the val set is correct). I just check your uploaded image. The prediction indexes of the uploaded image is incorrect, e.g. the road is misclassified as bicycle. That's really weird... I run the code on my computer. Both cbst and mrkld work fine. The only issue that may cause the problem may be the mapping function in the training process. To check if the mapped labels are correct, you may want to see the sampled patch's labels. That might be the problem to cause the error.

**Comment by prabhuteja12** [comment]
I've tried to look at individual predictions of the baseline network on CS Train set (first call to the `val` function). The predictions look correct (no mislabeling there).  

However when it runs as a single code (instead of me debugging in jupyter), same behavior resumes, and wrong labels are predicted,

**Comment by prabhuteja12** [comment]
> LRENT has different MRKLD and CBST. So it may be the library called in the dataset function is different than my settings. You may want to check it.

Can you clarify what you mean here? I understand different `Dataset` functions are used, is that where you think the difference might be from?

**Comment by yzou2** [comment]
Sorry for the late response, I am seeking a job right now and the response can be slow. The LRENT should load soft labels while CBST and MRKLD only require hard labels. That's the difference.  Can you just use the evaluation code to evaluate the intermediate models in the self-training? If these results are good. There might be some issues in the evaluation part of the self-training code.

**Comment by prabhuteja12** [comment]
Hi, I tried the `evaluate.py` to see if something changes, however its the same kind of label flipping that is happening. So there seems to be something happening in the training code itself.
Would it be possible for you to pull the public rep code and try executing the CBST part? Do you think there might be some difference between your personal copy of the code, and the one on git?

**Comment by yzou2** [comment]
Let me have a try.

**Comment by prabhuteja12** [comment]
Hi, any luck?

**Comment by yzou2** [comment]
I just downloaded the code and run cbst.sh on my desktop. The performance is good. Check the attached log.

[self_training_log.txt](https://github.com/yzou2/CRST/files/4535909/self_training_log.txt)

**Comment by prabhuteja12** [comment]
Hi, 
i downgraded my pytorch from 0.4.1  to 0.4.0 and it works. I wasn;t aware that pytorch had so many breaking changes in a minor revision. Do you know if this is true?

**Comment by yzou2** [comment]
I also had different performances across different pytorch versions, even in terms of the minor version change. Thanks for your feedback and glad to hear you solve the problem. I will update a notes for the repo for others to avoid such pitfalls. Thanks!

**Comment by prabhuteja12** [comment]
the easiest might be for you to share your environment.yml file so that recreating the exact conditions that worked for you the easiest. 

Thanks again!

**Comment by yzou2** [comment]
> the easiest might be for you to share your environment.yml file so that recreating the exact conditions that worked for you the easiest.
> 
> Thanks again!

Good point. I will do it when I am available.
